Unify widget state image names (disabled, pressed, hover, focused)

dragunoff opened the pull request:
This PR attempts to unify state suffixes used in image names across various widgets. Currently all of these suffixes are managed in each widget resulting in some inconsistencies in the naming and in the precedence of states. I looked at the widgets that have stateful names and extracted these 4 common states (in order of precedence):

1. `disabled`
1. `pressed`
1. `hover`
1. `focused`

The stateful name is composed of the base name and a suffix (e.g. `button-hover`).

I was considering making `highlighted` a state as well, but after thinking about it I realized it should stay as a variant of the base name (e.g. `button-highlighted-hover`).

This stared as a take over of #17612 but as I went down the rabbit hole I decided to make an overhaul. The goal of #17612 is still achieved though - I have added properties to configure the image collection for separators in drop-downs and also for arrows in drop-downs, scrollbars and production tabs (these used to be hard-coded to "scrollbar").

There is some more to do in this direction (with regard to checkboxes and scroll items) but I'll leave that to follow-ups to make reviewing easier. The core idea is in this PR.

dragunoff commented:
Updated to fix RA observer dropdown separator.

dragunoff commented:
Rebased.

dragunoff commented:
Updated to add a property for setting the image name for arrows and separators in drop-downs.

dragunoff commented:
Updated to fix a crash with the command bar in RA.

dragunoff commented:
Rebased after the merge of #17741

dragunoff commented:
Rebased because #17737 depends on it.

teinarss reviewed (dismissed):
works

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/CommandBarLogic.cs:
```diff
@@ -275,15 +275,16 @@ public override void Tick()
+			icon.GetImageName = () =>
+			{
+				var baseName = button.IsHighlighted() ? icon.ImageName + "-active" : icon.ImageName;
+				var hasBaseImage = ChromeProvider.GetImage(icon.ImageCollection, baseName) != null;
```
Querying these images 60+ times a second isn't great. We can pull some things outside `GetImageName` by e.g. having separate `hasImage` vs `hasActiveImage` and otherwise use `CachedTransform` to avoid doing work when nothing has changed.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/StanceSelectorLogic.cs:
```diff
@@ -48,15 +48,16 @@ public StanceSelectorLogic(Widget widget, World world)
+			icon.GetImageName = () =>
+			{
+				var baseName = button.IsHighlighted() ? icon.ImageName + "-active" : icon.ImageName;
+				var hasBaseImage = ChromeProvider.GetImage(icon.ImageCollection, baseName) != null;
```
Likewise here.

dragunoff committed:
Unify widget state image suffixes (disabled, pressed, hover, focus)

- Add a property for arrows image collection (in drop-downs, scrollbars
and production tabs)
- Add a property for separators image collection (in drop-downs)
- Add hover and disable states to the drop-down separator
- Unify button, textfield and checkbox state suffixes

dragunoff committed:
Refactor command bar buttons to use unified widget state names

dragunoff commented:
Updated to address @pchote's comment. I moved the whole command and stance logic refactor into a separate commit. But I could not figure out how to make it work with `CachedTransform`.

Also fixed a couple of yaml lines which I had overlooked.

dragunoff commented:
Rebased.

pchote commented on OpenRA.Mods.Common/Widgets/CheckboxWidget.cs:
```diff
@@ -54,11 +53,8 @@ public override void Draw()
-						Depressed && HasPressedState ? "checkbox-pressed" :
-						Ui.MouseOverWidget == this ? "checkbox-hover" :
-						"checkbox";
+			var baseName = IsHighlighted() ? "checkbox-highlighted" : "checkbox";
```
This leads to missing artwork / funky results if anyone ever highlights a checkbox. Can we please definitions for `checkbox-highlighted-hover`, `checkbox-highlighted-pressed`, and `checkbox-highlighted-disabled` to the chrome.yaml?

My testcase was to hardcode `IsHighlighted` and `IsDisabled` bindings to one of the settings checkboxes.

dragunoff replied:
The checkbox is one of the most complex input widgets and currently it's not handled very well in OpenRA. I have a follow up in the works that overhauls the states and naming conventions. For now I will add the missing definitions here so the current PR doesn't leave things in a weird state.

dragunoff replied:
Actually only TD has art for all of the highlighted states (both for button and checkbox). The rest of the mods just have a highlighted state and that's all. So I kept it simple in those mods by inheriting like this:
```yaml
checkbox-highlighted-hover:
	Inherits: checkbox-highlighted

checkbox-highlighted-disabled:
	Inherits: checkbox-disabled
```

pchote commented on OpenRA.Mods.Common/Widgets/DropDownButtonWidget.cs:
```diff
@@ -19,8 +20,10 @@ namespace OpenRA.Mods.Common.Widgets
-		public readonly string SeparatorImage = "separator";
+		public readonly string Arrows = "arrows";
+		public readonly string ArrowImage = "down";
+		public readonly string Separators = "separators";
```
IMO it would probably be tidier to instead have:
```
public readonly string Decorations = "dropdown-decorations";
public readonly string DecorationMarker = "marker";
public readonly string DecorationSeparator = "separator";
```
and
```
dropdown-decorations:
	Inherits: ^Chrome
	Regions:
		marker: 836, 17, 16, 16
		marker-pressed: 853, 17, 16, 16
		marker-disabled: 853, 17, 16, 16
		separator: 129, 2, 1, 19
		separator-hover: 161, 2, 1, 19
		separator-pressed: 129, 34, 1, 19
		separator-disabled: 161, 34, 1, 19
```

rather than mixing concepts between dropdowns and scrollpanels.

pchote replied:
I guess we could also have a single `button-decorations` image that covered both dropdowns and scrollpanels.

dragunoff replied:
This setup won't work when the marker and the separator come from different sprites. This is the case in every mod except TD and the reason why I made two separate properties for image collections.

So, perhaps go with something like:
```csharp
public readonly string Marker = "dropdown-markers";
public readonly string MarkerImage = "marker";
public readonly string Separator = "dropdown-separators";
public readonly string SeparatorImage = "separator";
```

pchote replied:
Hmm, i see. It might be simpler to just add the separator to glyphs.svg?

dragunoff replied:
I don't have a strong objection against that. We'd have to duplicate several small parts of the dialog graphics and that's somewhat fiddly and more difficult to maintain. But we'd get nicer semantics.

The separator is such an odd piece... I think it should be part of the marker graphic. But it's not because the marker is assembled from a separator and an arrow.

pchote replied:
I wouldn't object to combining them.

dragunoff replied:
I think that moving the separator to the glyph sprite (on its own or combined with the marker) is more trouble than it's worth :thinking: The separators are using the dialog art and moving them away would just incur maintenance cost. And furthermore they are an optional bit and mods can choose to ignore them and just use a marker image.

I suggest we leave it like this:
```csharp
public readonly string Decorations = "dropdown-decorations";
public readonly string DecorationMarker = "marker";
public readonly string Separators = "dropdown-separators";
public readonly string SeparatorImage = "separator";
```
Of course if you insist @pchote I will move them to the glyph sprite.

pchote replied:
Ok, if we can unify the other cases then i'm fine with leaving this one split.

pchote commented on OpenRA.Mods.Common/Widgets/ProductionTabsWidget.cs:
```diff
@@ -83,6 +83,7 @@ public class ProductionTabsWidget : Widget
 
 		public string Button = "button";
 		public string Background = "panel-black";
+		public string Arrows = "arrows";
```
And then here we could have
```
public readonly string Decorations = "scrollpanel-decorations";
public readonly string DecorationScrollLeft = "left";
public readonly string DecorationScrollRight = "right";
```

pchote commented on OpenRA.Mods.Common/Widgets/ScrollPanelWidget.cs:
```diff
@@ -48,6 +48,7 @@ public class ScrollPanelWidget : Widget
 		public string Background = "scrollpanel-bg";
 		public string ScrollBarBackground = "scrollpanel-bg";
 		public string Button = "scrollpanel-button";
```
and then
```
public readonly string Decorations = "scrollpanel-decorations";
public readonly string DecorationScrollUp = "up";
public readonly string DecorationScrollDown = "down";
```

pchote reviewed (commented):
The overall idea here LGTM, but IMO we are currently missing an opportunity to clarify the way we handle button decorations:

pchote commented:
Any progress here @dragunoff?

dragunoff commented:
@pchote I like the abstraction and semantics that you proposed. I just haven't gotten around to rework the PR. I will take a look at it very soon, it's at the top of my todos.

dragunoff commented:
Update: rebased and addressed comments from pchote's review (https://github.com/OpenRA/OpenRA/pull/17663#pullrequestreview-421485524).

dragunoff commented:
@pchote Is there anything more to do here? This is a dependency for #17737 and my main motivation :-)

pchote commented:
Sorry, this had dropped off my radar. I will try and get another look at this soon.

pchote commented:
Code changes LGTM.

One issue spotted while testing:
* Dropdown arrows in RA and TS use the grey disabled state when pressed rather than the white normal state (TD and D2k behave as expected)

Should be good to merge once this has been fixed.

dragunoff commented:
@pchote Good catch. This happened because in an earlier iteration I was using the same collection for both dropdowns and scrollbars (`arrows`). And scrollbars use the disabled image for the pressed state. And when I split them I just copy-pasted the `arrows` collection. 

Should I change the scrollbars to use the normal arrow as well? IMO it looks better and makes sense because the arrow is certainly not disabled while scrolling.

pchote commented:
> Should I change the scrollbars to use the normal arrow as well? IMO it looks better and makes sense because the arrow is certainly not disabled while scrolling.

Yeah, I think so too.

dragunoff commented:
Updated to address the issues with the pressed state of the arrows.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9083edc67c6ade0b4e9fba3338e0ae45365447a8

https://github.com/OpenRA/OpenRAModSDK/wiki/Update-notes:-Release-20200503-to-next-release
